File: frontend/lib/patient-profile-builder.ts
export type AllergySeverity = "mild" | "moderate" | "severe";

export interface ChronicCondition {
  icdCode: string;
  description: string;
}

export interface Allergy {
  name: string;
  severity: AllergySeverity;
}

export interface EmergencyContact {
  name: string;
  phone: string;
}

export interface InsuranceDetails {
  provider: string;
  policyNumber: string;
  expiry: string;
}

export interface PatientProfile {
  id: string;
  name: string;
  age: number;
  gender: string;
  email: string;
  phone: string;
  createdByDoctorEmail: string;
  chronicConditions: ChronicCondition[];
  allergies: Allergy[];
  emergencyContact?: EmergencyContact;
  insuranceDetails?: InsuranceDetails;
  createdAt: string;
}

export class PatientProfileBuilder {
  private static readonly KNOWN_ICD_CODES = new Set([
    "E11.9",
    "I10",
    "J45.909",
    "E78.5",
    "M54.5",
  ]);

  private profile: Partial<PatientProfile> = {
    chronicConditions: [],
    allergies: [],
  };

  setCore(
    name: string,
    age: number,
    gender: string,
    email: string,
    phone: string,
  ): this {
    this.profile.name = name.trim();
    this.profile.age = age;
    this.profile.gender = gender.trim();
    this.profile.email = email.trim();
    this.profile.phone = phone.trim();
    return this;
  }

  setCreatedByDoctorEmail(doctorEmail: string): this {
    this.profile.createdByDoctorEmail = doctorEmail.trim();
    return this;
  }

  addChronicCondition(icdCode: string, description: string): this {
    const normalizedCode = icdCode.trim();
    const normalizedDescription = description.trim();
    if (!normalizedCode || !normalizedDescription) {
      return this;
    }

    this.profile.chronicConditions?.push({
      icdCode: normalizedCode,
      description: normalizedDescription,
    });
    return this;
  }

  addAllergy(substance: string, severity: AllergySeverity): this {
    const trimmedSubstance = substance.trim();
    if (!trimmedSubstance) {
      return this;
    }

    this.profile.allergies?.push({ name: trimmedSubstance, severity });
    return this;
  }

  setEmergencyContact(name: string, phone: string): this {
    const trimmedName = name.trim();
    const trimmedPhone = phone.trim();
    if (!trimmedName || !trimmedPhone) {
      return this;
    }

    this.profile.emergencyContact = {
      name: trimmedName,
      phone: trimmedPhone,
    };
    return this;
  }

  setInsuranceDetails(details: InsuranceDetails): this {
    const provider = details.provider.trim();
    const policyNumber = details.policyNumber.trim();
    const expiry = details.expiry.trim();

    if (!provider || !policyNumber || !expiry) {
      return this;
    }

    this.profile.insuranceDetails = {
      provider,
      policyNumber,
      expiry,
    };

    return this;
  }

  build(): PatientProfile {
    if (!this.profile.name) {
      throw new Error("Patient name is required.");
    }
    if (!this.profile.age || this.profile.age <= 0) {
      throw new Error("Patient age must be a positive number.");
    }
    if (!this.profile.gender) {
      throw new Error("Patient gender is required.");
    }
    if (!this.profile.email || !this.profile.email.includes("@")) {
      throw new Error("A valid patient email is required.");
    }
    if (!this.profile.phone) {
      throw new Error("Patient phone is required.");
    }
    if (!this.profile.createdByDoctorEmail) {
      throw new Error("Creating doctor context is required.");
    }

    for (const condition of this.profile.chronicConditions ?? []) {
      if (!PatientProfileBuilder.KNOWN_ICD_CODES.has(condition.icdCode)) {
        throw new Error(`Invalid ICD code for chronic condition: ${condition.icdCode}`);
      }
    }

    const contact = this.profile.emergencyContact;
    if (contact && (!contact.name?.trim() || !contact.phone?.trim())) {
      throw new Error(
        "Emergency contact is partially filled: both name and phone are required.",
      );
    }

    const insurance = this.profile.insuranceDetails;
    if (
      insurance &&
      (!insurance.provider?.trim() ||
        !insurance.policyNumber?.trim() ||
        !insurance.expiry)
    ) {
      throw new Error(
        "Insurance details are partially filled: provider, policy number, and expiry are required.",
      );
    }

    return {
      id: this.profile.id ?? `patient-${Date.now()}`,
      name: this.profile.name,
      age: this.profile.age,
      gender: this.profile.gender,
      email: this.profile.email,
      phone: this.profile.phone,
      createdByDoctorEmail: this.profile.createdByDoctorEmail,
      chronicConditions: this.profile.chronicConditions ?? [],
      allergies: this.profile.allergies ?? [],
      ...(this.profile.emergencyContact
        ? { emergencyContact: this.profile.emergencyContact }
        : {}),
      ...(this.profile.insuranceDetails
        ? { insuranceDetails: this.profile.insuranceDetails }
        : {}),
      createdAt: new Date().toISOString(),
    };
  }
}

File: backend/java/src/main/java/com/scribehealth/model/PatientProfileBuilder.java
package com.scribehealth.model;

import java.time.Instant;
import java.time.LocalDate;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public class PatientProfileBuilder {

    private static final Set<String> KNOWN_ICD_CODES = Set.of(
            "E11.9",  // Type 2 diabetes mellitus without complications
            "I10",    // Essential (primary) hypertension
            "J45.909",// Unspecified asthma, uncomplicated
            "E78.5",  // Hyperlipidemia, unspecified
            "M54.5"   // Low back pain
    );

    private String name;
    private Integer age;
    private String gender;
    private String email;
    private String phone;

    private String createdByDoctorEmail;

    private final List<PatientProfile.ChronicCondition> chronicConditions = new ArrayList<>();
    private final List<PatientProfile.AllergyEntry> allergies = new ArrayList<>();

    private String emergencyContactName;
    private String emergencyContactPhone;

    private String insuranceProvider;
    private String insurancePolicyNumber;
    private String insuranceExpiry;

    public PatientProfileBuilder setCore(String name, int age, String gender, String email, String phone) {
        this.name = normalize(name);
        this.age = age;
        this.gender = normalize(gender);
        this.email = normalize(email);
        this.phone = normalize(phone);
        return this;
    }

    public PatientProfileBuilder setCreatedByDoctorEmail(String doctorEmail) {
        this.createdByDoctorEmail = normalize(doctorEmail);
        return this;
    }

    public PatientProfileBuilder addChronicCondition(String icdCode, String description) {
        this.chronicConditions.add(
                new PatientProfile.ChronicCondition(normalize(icdCode), normalize(description))
        );
        return this;
    }

    public PatientProfileBuilder addAllergy(String substance, String severity) {
        String normalizedSubstance = normalize(substance);
        String normalizedSeverity = normalize(severity);
        PatientProfile.AllergySeverity parsedSeverity;
        try {
            parsedSeverity = PatientProfile.AllergySeverity.valueOf(normalizedSeverity.toUpperCase());
        } catch (IllegalArgumentException ex) {
            throw new IllegalArgumentException("Allergy severity must be one of: mild, moderate, severe");
        }

        this.allergies.add(new PatientProfile.AllergyEntry(normalizedSubstance, parsedSeverity));
        return this;
    }

    public PatientProfileBuilder setEmergencyContact(String name, String phone) {
        this.emergencyContactName = normalize(name);
        this.emergencyContactPhone = normalize(phone);
        return this;
    }

    public PatientProfileBuilder setInsurance(String provider, String policyNumber, String expiry) {
        this.insuranceProvider = normalize(provider);
        this.insurancePolicyNumber = normalize(policyNumber);
        this.insuranceExpiry = normalize(expiry);
        return this;
    }

    public PatientProfile build() {
        validateCore();
        validateDoctorScope();
        validateChronicConditions();
        validateEmergencyContact();
        validateInsurance();

        PatientProfile profile = new PatientProfile();
        profile.setName(name);
        profile.setAge(age);
        profile.setGender(gender);
        profile.setEmail(email);
        profile.setPhone(phone);
        profile.setCreatedByDoctorEmail(createdByDoctorEmail);
        profile.setCreatedAt(Instant.now());
        profile.setChronicConditions(new ArrayList<>(chronicConditions));
        profile.setAllergies(new ArrayList<>(allergies));

        if (hasText(emergencyContactName) && hasText(emergencyContactPhone)) {
            profile.setEmergencyContact(new PatientProfile.EmergencyContact(emergencyContactName, emergencyContactPhone));
        }

        if (hasText(insuranceProvider) && hasText(insurancePolicyNumber) && hasText(insuranceExpiry)) {
            profile.setInsuranceDetails(
                    new PatientProfile.InsuranceDetails(
                            insuranceProvider,
                            insurancePolicyNumber,
                            LocalDate.parse(insuranceExpiry)
                    )
            );
        }

        return profile;
    }

    private void validateCore() {
        if (!hasText(name)) {
            throw new IllegalArgumentException("Patient name is required");
        }
        if (age == null || age <= 0) {
            throw new IllegalArgumentException("Patient age must be a positive number");
        }
        if (!hasText(gender)) {
            throw new IllegalArgumentException("Patient gender is required");
        }
        if (!hasText(email) || !email.contains("@")) {
            throw new IllegalArgumentException("A valid patient email is required");
        }
        if (!hasText(phone)) {
            throw new IllegalArgumentException("Patient phone is required");
        }
    }

    private void validateDoctorScope() {
        if (!hasText(createdByDoctorEmail)) {
            throw new IllegalArgumentException("Creating doctor context is required");
        }
    }

    private void validateChronicConditions() {
        for (PatientProfile.ChronicCondition condition : chronicConditions) {
            if (!hasText(condition.getIcdCode()) || !KNOWN_ICD_CODES.contains(condition.getIcdCode())) {
                throw new IllegalArgumentException(
                        "Invalid ICD code for chronic condition: " + condition.getIcdCode()
                );
            }
            if (!hasText(condition.getDescription())) {
                throw new IllegalArgumentException("Chronic condition description is required");
            }
        }
    }

    private void validateEmergencyContact() {
        boolean hasName = hasText(emergencyContactName);
        boolean hasPhone = hasText(emergencyContactPhone);

        if (hasName ^ hasPhone) {
            throw new IllegalArgumentException(
                    "Emergency contact is partially filled: both name and phone are required"
            );
        }
    }

    private void validateInsurance() {
        boolean hasProvider = hasText(insuranceProvider);
        boolean hasPolicyNumber = hasText(insurancePolicyNumber);
        boolean hasExpiry = hasText(insuranceExpiry);

        if (hasProvider || hasPolicyNumber || hasExpiry) {
            if (!(hasProvider && hasPolicyNumber && hasExpiry)) {
                throw new IllegalArgumentException(
                        "Insurance details are partially filled: provider, policyNumber, and expiry are all required"
                );
            }

            try {
                LocalDate.parse(insuranceExpiry);
            } catch (DateTimeParseException ex) {
                throw new IllegalArgumentException("Insurance expiry must be in ISO date format: YYYY-MM-DD");
            }
        }
    }

    private String normalize(String value) {
        return value == null ? null : value.trim();
    }

    private boolean hasText(String value) {
        return value != null && !value.isBlank();
    }
}